### DISCUSSION TITLE
Allow plugins to change mayPickup() logic.

### DISCUSSION THREAD

**Discussion by WillyTheWorm** [PaperMC/Paper#10551] [Ideas]
### Is your feature request related to a problem?

When using `OnCraftingEvent` or similar, you can change crafting results using `setResult()`, but the player is unable to remove the item from the crafting GUI.

### Describe the solution you'd like.

Allow plugins to change `mayPickup()` logic, so it is possible to change crafting outputs _and_ allow the player to remove them from the GUI.

### Describe alternatives you've considered.

A workaround is to just create a recipe, but this can be tedious, and there are limitations that `mayPickup()` would fix.

**Comment by Owen1212055** [comment]
Can you please explain WHICH event exactly you want this behavior with? I am struggling to see events that don't allow you to pickup the item when using #setResult().

Or are you referring to (Some Inventory)#setResult?
Would appreciate if more info could be given here.

**Comment by lynxplay** [comment]
The one place where this is important is the `SmithingMenu#mayPickup(Player,boolean)` method, which reads as follows:
```java
@Override
protected boolean mayPickup(Player player, boolean present) {
    return this.selectedRecipe != null && this.selectedRecipe.matches(this.inputSlots, this.level);
}
```

Issue here being that even after modification of the result slot, this method will prevent grabbing the item if no recipe is available matching it.
